src/rider/main/kotlin/previewer/AvaloniaMessageMouseListener.kt

```diff
@@ -0,0 +1,149 @@
+            coordinates.second,
+            e.avaloniaMouseButton())
+        avaloniaInputEventSignal.fire(message)
+    }
+
+    override fun mouseWheelMoved(e: MouseWheelEvent?) {
+        e ?: return
+        if (e.scrollType != MouseWheelEvent.WHEEL_UNIT_SCROLL) return
+        val coordinates = e.relCoordinatesOrNull()?: return
+        val message = ScrollEventMessage(
+            e.avaloniaModifiers(),
+            coordinates.first,
+            coordinates.second,
+            0.0,
+            -e.preciseUnitsToScroll())
+        avaloniaInputEventSignal.fire(message)
+    }
+
+    override fun mouseMoved(e: MouseEvent?) {
+        sendPointerMovedEventMessage(e)
+    }
+
+    override fun mouseDragged(e: MouseEvent?) {
+        sendPointerMovedEventMessage(e)
+    }
+
+    private fun sendPointerMovedEventMessage(e: MouseEvent?) {
+        e ?: return
+        val coordinates = e.relCoordinatesOrNull()?: return
+        val message = PointerMovedEventMessage(
+            e.avaloniaModifiers(),
+            coordinates.first,
+            coordinates.second)
+        avaloniaInputEventSignal.fire(message)
+    }
+
+    private fun MouseWheelEvent.preciseUnitsToScroll(): Double {
+        return this.preciseWheelRotation * this.scrollAmount
+    }
+
+    private fun MouseEvent.relCoordinatesOrNull(): Pair<Double, Double>? {
+        frameView.icon ?: return null
+        val relFirst = frameView.relFirstOrNull(this.x) ?: return null
+        val relSecond = frameView.relSecondOrNull(this.y) ?: return null
+        return Pair(relFirst, relSecond)
+    }
+
+    private fun MouseEvent.avaloniaModifiers(): Array<Int> {
+        val result = mutableListOf<Int>()
+
+        val m = this.modifiersEx
+        if ((m and MouseEvent.ALT_DOWN_MASK) != 0)
+            result.add(InputModifiers.Alt.ordinal)
+        if ((m and MouseEvent.CTRL_DOWN_MASK) != 0)
+            result.add(InputModifiers.Control.ordinal)
+        if ((m and MouseEvent.SHIFT_DOWN_MASK) != 0)
+            result.add(InputModifiers.Shift.ordinal)
+        if ((m and MouseEvent.META_DOWN_MASK) != 0)
+            result.add(InputModifiers.Windows.ordinal)
+        if ((m and MouseEvent.BUTTON1_DOWN_MASK) != 0)
+            result.add(InputModifiers.LeftMouseButton.ordinal)
+        if ((m and MouseEvent.BUTTON2_DOWN_MASK) != 0)
+            result.add(InputModifiers.MiddleMouseButton.ordinal)
+        if ((m and MouseEvent.BUTTON3_DOWN_MASK) != 0)
+            result.add(InputModifiers.RightMouseButton.ordinal)
+
+        return result.toTypedArray()
+    }
+
+    private fun MouseEvent.avaloniaMouseButton(): Int {
+        return when (this.button) {
+            MouseEvent.BUTTON1 -> MouseButton.Left.ordinal
+            MouseEvent.BUTTON2 -> MouseButton.Middle.ordinal
+            MouseEvent.BUTTON3 -> MouseButton.Right.ordinal
+            else -> MouseButton.None.ordinal
+        }
+    }
+
+    private fun JLabel.relFirstOrNull(first: Int): Double? {
```

Comment:
It is unclear what this method is doing. Is it just `relativeXCoordinate`?

The same about `relSecondOrNull`, is it the same as `relativeYCoordinate`?

Reply:
I as a result have refused these methods